src/components/implementation/no_interface/new_boot/boot_deps.h

```diff
@@ -0,0 +1,147 @@
+#include <print.h>
+
+#undef assert
+#define assert(node) do { if (unlikely(!(node))) { debug_print("assert error in @ "); *((int *)0) = 0;} } while(0)
+
+#include <cos_alloc.h>
+#include <cobj_format.h>
+#include <cos_types.h>
+
+/*the booter uses this to keep track of each comp*/
+struct comp_cap_info {
+	struct cos_compinfo cos_compinfo;
+	vaddr_t addr_start;
+	vaddr_t vaddr_mapped_in_booter;
+	vaddr_t upcall_entry;
+};
+struct comp_cap_info comp_cap_info[MAX_NUM_SPDS+1];
+
+struct cos_compinfo boot_info;
+
+thdcap_t schedule[MAX_NUM_SPDS+1];
+unsigned int cur_sched;
+
+static int
+boot_comp_map_memory(struct cobj_header *h, spdid_t spdid, pgtblcap_t pt)
+{
+	int i, j;
+	int flag;
+	vaddr_t dest_daddr, prev_map = 0;
+	int tot = 0, n_pte = 1;	
+	
+	struct cobj_sect *sect = cobj_sect_get(h, 0);
+	
+	/*Expand Page table, could do this faster*/
+	for(j = 0; j < (int)h->nsect; j++){
+		tot += cobj_sect_size(h, j);
+	}
+	printc("tot: %d\n", tot);
+	if (tot > SERVICE_SIZE) {
+		n_pte = tot / SERVICE_SIZE;
+		if(tot % SERVICE_SIZE) n_pte++;
+	}	
+	for(j = 0; j < n_pte; j++){
+		if (!__bump_mem_expand_range(&boot_info, pt, sect->vaddr, SERVICE_SIZE)) BUG();
+	}
+
+	/* We'll map the component into booter's heap. */
+	comp_cap_info[spdid].vaddr_mapped_in_booter = (vaddr_t)cos_get_heap_ptr();
+	
+	for (i = 0 ; i < h->nsect ; i++) {
+		int left;
+
+		sect = cobj_sect_get(h, i);
+		flag = MAPPING_RW;
+		if (sect->flags & COBJ_SECT_KMEM) {
+			flag |= MAPPING_KMEM;
+		}
+
+		dest_daddr = sect->vaddr;
+		left       = cobj_sect_size(h, i);
+		
+		/* previous section overlaps with this one, don't remap! */
+		if (round_to_page(dest_daddr) == prev_map) {
+			left -= (prev_map + PAGE_SIZE - dest_daddr);
+			dest_daddr = prev_map + PAGE_SIZE;
+		}
+
+		while (left > 0) {
+			vaddr_t addr = cos_page_bump_alloc(&boot_info);
+			assert(addr);
+			
+			if (cos_mem_alias_at(&comp_cap_info[spdid].cos_compinfo, dest_daddr, &boot_info, addr)) BUG();
```

Comment:
It looks like the mappings are always contiguous (one after the other in increasing pages).  If that's the case, I feel like we can get rid of the `expand_range`, and instead just use `cos_mem_alias` (after initializing the heap pointer to the proper starting address).  I know you explained to me before why this is required here, but I can't reason it out now.